proposals/aws-licensing-integration.md

```diff
@@ -0,0 +1,88 @@
+# AWS licensing integration
+
+This specification documents the changes we would need in Fleet to integrate with AWS licensing.
+
+## Why
+
+> AWS requires us to offer a pricing option native to AWS Marketplace ~90 days after launch. To achieve this, we need to go through various technical reviews and build an integration to their pricing system.
+
+## Documentation
+
+- https://docs.aws.amazon.com/marketplace/latest/userguide/container-license-manager-integration.html
+- https://aws.amazon.com/blogs/awsmarketplace/aws-license-manager-integration-aws-marketplace-ami-containers-contract-pricing/
+
+## Successful integration?
+
+How does AWS determine whether we integrated to their licensing system the way they expect? (TBD!)
+
+## Product requirements
+
+Following are the requirements from the [epic issue](https://github.com/fleetdm/fleet/issues/8004).
+1. Use the contract-based pricing option w/ our current pricing ($1/host/month)
+2. Enable private offers.
+
+We believe that once we implement requirement #1 in Fleet, requirement #2 only needs a AWS/cloud configuration (it's just a different agreed upon rate than the listed one).
+
+In other words:
+- #1 requires code changes in Fleet.
+- #2 requires #1, and,
+- #2 requires no code changes in Fleet.
+
+## How
+
+Assumption: To integrate with the AWS licensing model, the application (Fleet) must use the AWS License Manager API (instead of using own/current licensing functionality).
+
+## Some TBDs around requirements
+
+- What do we want Fleet to do when a AWS license expires? (Currently we don't restrict usage much when a license expires.)
+- Do we want to delete hosts if the number of hosts is reduced in a purchased license? (not sure how feasible of a scenario this is.)
+
+### Must have
+
+Limit the number of hosts allowed to enroll to Fleet depending on the license "resource unit" (host count) defined in the purchased AWS license.
+
+### Nice to have (can be done in a later iteration)
+
+The following features wouldn't block the AWS licensing integration (thus they are nice to have):
+
+- Show license about to expire banner on the UI.
+- Show banner when a user has reached the maximum number of hosts allowed in their purchased license.
```

Comment:
> - Show license about to expire banner on the UI.

This is the case for current Fleet licensing. What I meant here is that we'll need the same for this new AWS licensing.

Yes. There are some small bugs I'll be opening in separate issues.

> - Show banner when a user has reached the maximum number of hosts allowed in their purchased license.

Just tested this and looks like we don't notify the user in the UI when they've reached the host limit (you can only see the "host limit" errors in the fleet logs and by noticing that your hosts cannot enroll)

Reply:
If we're not doing it in the current fleet UI, when let's not do it for this. 

So based on that, it sounds like we should use the same banner for when the license is about to expire and/or expired. And we don't have any UI when you're over your host count (but you get logs) and we refuse to enroll new hosts.

Reply:
In order to maintain the same functionality, we should "translate" the AWS license to something that looks like one of ours, because that's used in some APIs and the frontend reasons about it based on that.